### ISSUE TITLE
UWP Support Tracker

### ISSUE THREAD

**Issue by flibitijibibo** [flibitijibibo/SDL2-CS#114]
This is a thread that tracks support for UWP 6.0, targeting Windows 10 16299 (Fall Creator's Update).

This is a follow-up to #93.

**Current Status:** Code compiles and runs on desktop and Xbox with SDL 2.0.8 and the latest version of FAudio.

**Current Third-Party Bugs:**
- SDL:
    - [WASAPI Bug](https://bugzilla.libsdl.org/show_bug.cgi?id=4224)
    - [SDL_image](https://bugzilla.libsdl.org/show_bug.cgi?id=3965)/[SDL_ttf](https://bugzilla.libsdl.org/show_bug.cgi?id=3966) Project Files Missing
- [ANGLE WindowsStore](https://github.com/microsoft/angle):
    - All kinds of dumb stuff, just look at [FNA's branch](https://github.com/FNA-XNA/angle) instead.
- MSBuild
    - Dumb bug where obj/ folders collide, [worked around in FNA](https://github.com/FNA-XNA/FNA/commit/07ce777af8b8fdee4268c190bc73aba2ec5a24ab)

**Code to Port to UWP:**
```
    [STAThread]
    static void Main(string[] args)
#if WINDOWS_UWP
    {
        realArgs = args;
        SDL2.SDL.SDL_SetHint("SDL_WINRT_HANDLE_BACK_BUTTON", "1");
        SDL2.SDL.SDL_WinRT_mainFunction mainFunction = FakeMain;
        SDL2.SDL.SDL_WinRTRunApp(mainFunction, IntPtr.Zero);
    }

    static string[] realArgs;
    static int FakeMain(int argc, IntPtr[] argv)
    {
        RealMain(realArgs);
        return 0;
    }

    static void RealMain(string[] args)
#endif
    {
        // blah blah blah
    }
```

For single-assembly portability, remove the `#if` check and go straight to RealMain if `!SDL2.SDL.SDL_GetPlatform().Equals("WinRT")`.

Slack channel for anyone that wants to help fix the third-party bugs and/or put UWP builds together: [Link](https://join.slack.com/t/fna-xbone/shared_invite/enQtMzA1MjMxNzQ2NDgxLWM4NTk2OTcxYWUyMmNhZjg0NmMyMWQxOGVkZmY2MGQ3YWNhYjQ1M2NmMWM0YTc3NmIzNjVhMDY3YzMzNzc3OTY)

**Comment by flibitijibibo** [comment]
Here's the first screenshot of FNA running on Windows 10 UWP:

<img width="720" alt="screen shot 2017-10-17 at 4 42 03 pm" src="https://user-images.githubusercontent.com/1396814/31688467-6f09d9a2-b35a-11e7-8f2f-32cd4d8420cb.png">

Initial Notes:
- SDL2 requires a quick patch to boot in a C# context; simply add `Windows::Foundation::Initialize(RO_INIT_MULTITHREADED)` to the beginning of SDL_WinRTRunApp(). Going to talk to David Ludwig about this.
- UWP ANGLE on NuGet is either super old or _severely_ limits your GLES capabilities for reasons unclear to me. Creating an ES3 context fails, but ES2 is okay, on a platform that requires D3D11 anyway. No clue what that's about.
- UWP is a fat liar about supporting Standard 2.0 with regard to UnmanagedType.CustomMarshaler. It compiles, but fails at runtime with a PlatformNotSupportedException. So much for it definitely and totally using .NET Core behind the scenes (trust us)!
- You can use the UWP Blank App template, strip out the XAML source files, and add your own source instead. FNA.Core.csproj (or SDL2-CS.Core.csproj) can be referenced by just adding it to the sln and referencing it in the main project, just like the old project files.
- Here is the code to make your existing FNA/SDL2# application work:

```
    [STAThread]
    static void Main(string[] args)
#if WINDOWS_UWP
    {
        Environment.SetEnvironmentVariable("FNA_OPENGL_FORCE_ES3", "1");
        Environment.SetEnvironmentVariable("SDL_OPENGL_ES_DRIVER", "1");
        Environment.SetEnvironmentVariable("FNA_AUDIO_DISABLE_SOUND", "1");
        SDL2.SDL.SDL_WinRT_mainFunction mainFunction = FakeMain;
        SDL2.SDL.SDL_WinRTRunApp(mainFunction, IntPtr.Zero);
    }

    static int FakeMain(int argc, IntPtr[] argv)
    {
        RealMain(new string[0]);
        return 0;
    }

    static void RealMain(string[] args)
#endif
    {
        // blah blah blah
    }
```

**Comment by flibitijibibo** [comment]
Asked the MSDN forums about the Xbone issue:

https://social.msdn.microsoft.com/Forums/windowsapps/en-US/6c852d53-e2f6-43e8-befc-c3075ab59fd3/rs31709xbox-one-c-fall-creators-update-16299-support?forum=wpdevelop

If someone tries this on Xbone and succeeds, DO NOT WAIT and let me know that it's just me. Until this can be resolved I basically can't work on this at all.

**Comment by flibitijibibo** [comment]
I just did a giant commit that moves from LPUtf8StrMarshaler over to our own UTF8 converters:

https://github.com/flibitijibibo/SDL2-CS/commit/795db4433d0794731566c5790d43da36c33a1a66

Upsides:
- UWP 6.0 compliant
- No malloc/free for input strings
- Less crap going on underneath the P/Invoke calls
- Less code for the UTF8 conversions themselves

Downsides:
- Holy INTERNAL_ functions, Batman
- Converting return strings still kinda sucks
- SDL_ShowMessageBox is still one scary mf

Some day when .NET 4.0 doesn't work on Windows anymore we can probably move to LPUtf8String, but that'll be ages off. Point being, this seemed like the best approach since I find it unlikely that Microsoft will ever care about UnmanagedType.CustomMarshaler for such a closed environment (that includes Windows 10).

Speaking of Microsoft, it seems someone forgot to boot an Xbox at their offices, because that bug I hit was known and is """Under Consideration""":

https://developercommunity.visualstudio.com/content/problem/130393/net-uwp-applications-with-a-targetplatformminversi.html

So Xbone support is definitely dead until they fix that. UWP on Windows, however, should work just fine now!

**Comment by flibitijibibo** [comment]
SDL now supports C# UWP out of the box:

https://hg.libsdl.org/SDL/rev/bd1495ebc979

Once Microsoft fixes Xbox One support, SDL2# UWP is done.

**Comment by flibitijibibo** [comment]
The latest Xbox update (still in preview) fixes SDL2# deployment. AFAIK, this is now done.

As a bonus, if you're using ANGLE and want to sidestep Xbone D3D only supporting 10_0 and not 10_1 (wtf) you can simply patch it to let you pick GLES 3. Just don't use stuff like cubemap arrays and you should be fine:

https://gist.github.com/flibitijibibo/9bc48cbe792f15d9f4ca4f344c085d3a

**Comment by flibitijibibo** [comment]
Found a bug in SDL-UWP's XInput implementation, in case anyone else runs into this:

https://bugzilla.libsdl.org/show_bug.cgi?id=3960

**Comment by flibitijibibo** [comment]
The latest Xbox dashboard update now works, so for SDL2# this issue is done. But let's also use this issue to update all the SDL libs for UWP...

SDL Hg boots but still needs [Bugzilla 3960](https://bugzilla.libsdl.org/show_bug.cgi?id=3960) for controllers. We also need to build, at minimum, [SDL_image](https://hg.libsdl.org/SDL_image/) for FNA. However, _image and _ttf only have Win8 projects and not UWP. Interestingly, _mixer does have UWP projects. So that's... cool...

CC @oldteen

**Comment by flibitijibibo** [comment]
Bugs filed for SDL_image and SDL_ttf UWP project files:

https://bugzilla.libsdl.org/show_bug.cgi?id=3965

https://bugzilla.libsdl.org/show_bug.cgi?id=3966

**Comment by kiates** [comment]
After getting things setup for UWP on my end I decided to see how far I could get on the Xbone. 

- Like you mentioned it wouldn't configure the graphic device until I forced ES V2 (I'm also using the ANGLE.WindowsStore nuget package right now).
- When I configure the AudioManager, it had trouble with the path to my sound/wave banks, etc.  It was using a current directory of C:\Windows\SysWOW64, so I hacked around it by pulling the directory of the assembly for absolute path generation.

After getting through those, no sprites would draw, but the background was the right color ;)  Even the simple splash screen logo I have.  I haven't dug into it any further yet. Is this by chance related to the necessary rebuilds of SDL_image?  I have VS 2015 still installed so could look into setting up the projects similar to the SDL. As much of a newb I am to FNA, am even more of one to SDL. :)

**Comment by flibitijibibo** [comment]
The lack of textures is due to GLES2. To use GLES3, ANGLE requires building from source. Grab this...

https://github.com/microsoft/angle

... then patch it with this...

https://gist.github.com/flibitijibibo/9bc48cbe792f15d9f4ca4f344c085d3a

As for the banks, the streams are opened with TitleContainer so I believe you can fix this by sending relative paths instead of absolute (which is what it sounds like was happening...?).

**Comment by kiates** [comment]
Thanks! That did the trick. I guess if I had tried running it again on Windows 10 after my hack, I would have seen the same effect.

My paths for the xact wave/sound banks were relative to begin with, so not sure what the culprit is there.

I'm trying to patch SDL with the game controller fix you proposed (VS 2015 keeps crashing now when loading the solution for some reason... arggghhhh).

**Comment by trufun202** [comment]
Leaving a comment, as requested by Ethan.  :)

**Comment by michaelarts** [comment]
Also leaving a comment to show my interest in this!

**Comment by michaelarts** [comment]
I'd love to help on the documentation side of things... but I won't be able to start on this until late next month. Once I'm free I'll take a look and see where I can assist! 

In regards to the room, I actually don't have any of those. Does anyone use Gitter?

**Comment by trufun202** [comment]
Slack gets my vote, but I'm up for whatever.

**Comment by flibitijibibo** [comment]
[Slack channel has been created.](https://join.slack.com/t/fna-xbone/shared_invite/enQtMzA1MjMxNzQ2NDgxLWM4NTk2OTcxYWUyMmNhZjg0NmMyMWQxOGVkZmY2MGQ3YWNhYjQ1M2NmMWM0YTc3NmIzNjVhMDY3YzMzNzc3OTY)

There are Electron desktop apps as well as the web version, either one works reasonably well and this is what I use for some other console dev channels, so this turns out to be pretty simple if you end up joining those channels as well (if you have Switch SDK access, ping me).

CC: @kiates @trufun202 @michaelarts

**Comment by michaelarts** [comment]
Sweet, that sounds easy enough... I can just use the web version! Also RE: Switch, I'm a licensed Nintendo developer but I've been unable to get access to the Switch SDK. Maybe I don't have enough reputation or something but they seem pretty guarded about it. If anyone has a contact at Nintendo or advice for that let me know lol.

**Comment by flibitijibibo** [comment]
The last known SDL bug for FNA Xbone has been fixed:

https://hg.libsdl.org/SDL/rev/1a033404c88b

Note that we still need UWP projects for SDL_image (and maybe SDL_ttf if you want), but that's just project files and not real code.

2.0.8 is scheduled to be out soon, so if you use SDL_image in your game, consider throwing together those project files!

**Comment by michaelarts** [comment]
Sweet! Great work. We're about to finish our new title then I'll switch gears towards porting to Xbox One with this. How is the audio going?

I've been working on a class that interfaces with XSAPI, which is needed to program things like achievements. If you'd like, I can contribute this to FNA somehow.... not sure if you'd want to include it in the repository but maybe it can be something that's linked in the wiki.

**Comment by michaelarts** [comment]
Also worth noting: my class implements basic saving and loading with the ConnectedStorage API... which is the required way to handle storage for ID@Xbox developers.

**Comment by flibitijibibo** [comment]
If it maps well to the old GamerServices/Net APIs then it may be worth making an FNA.XSAPI that works similar to FNA.Steamworks:

https://github.com/flibitijibibo/FNA.Steamworks

For saving/loading right now we're using Windows::Storage:

https://hg.libsdl.org/SDL/file/1a033404c88b/src/filesystem/winrt/SDL_sysfilesystem.cpp

I dunno how easy it'd be to make an ID@Xbox path but that's currently what we've got. If it maps to Xna.Storage that may be a route to take as well; FNA.Steamworks could have added a system to allow for custom StorageDevices to replace the default in FNA but it didn't seem all that necessary due to Steam autocloud existing.

**Comment by michaelarts** [comment]
Hmm.... yeah I'd have to look at the old GamerServices stuff but I think quite a few of the functions would map to that. If worse comes to worse we can just link to my code somewhere, it's nothing crazy but there's some essential stuff in here that people will need to pass cert.

I'll keep an eye on this thread; please keep me updated on when the audio is complete! Once that's in I think I'll have everything I need to dive in with my title.

**Comment by flibitijibibo** [comment]
SDL 2.0.8 and SDL_image 2.0.3 have been released:

https://www.libsdl.org/download-2.0.php

https://www.libsdl.org/projects/SDL_image/

This should have 100% of FNA's Xbox One changes integrated and should work out of the box. However, we still don't have VS2015 projects for SDL_image, so someone will need to generate those projects to build the UWP builds.

**Comment by flibitijibibo** [comment]
Remember this old thing?

I just made a fork of Microsoft's ANGLE repository to fix the build, fix a bug that they _still_ have not properly addressed, and adds our little secret Xbox hack:

https://github.com/FNA-XNA/angle

Build with VS2017 and the Release|x64 version should run on Xbox without any problems.

**Comment by flibitijibibo** [comment]
Bumping this with one last bug: This patch is needed to make WASAPI not totally messed up on Xbox:

https://bugzilla.libsdl.org/show_bug.cgi?id=4224

**Comment by flibitijibibo** [comment]
Closing this in favor of the [FNA documentation](https://github.com/FNA-XNA/FNA/wiki/Appendix-A:-FNA-on-Consoles). SDL2# is known to work on Xbox One, it just needs the help of native library patches that we have to deal with separately.

**Comment by flibitijibibo** [comment]
For those that joined the Slack channel, FNA now has a Discord server that includes a channel for Xbone development as well as many other topics: https://discord.gg/2Gg8zju
